### ISSUE TITLE
Add textures to transforms with shaders

### ISSUE THREAD

**Issue by shawna-p** [renpy/renpy#3862]
I am currently unaware of any easy ways of adding a texture to a transform which contains a shader. Consider, for example, the following situation:

- I am displaying my sprites during a forest scene. I would like the colouring of the sprites to look as though they are shaded by bright light coming through window blinds.
- I have created a shader, which requires two images/textures. The first is the sprite which is being displayed. The second is a control image of window blinds shading with black for shadows and white for light.
- By looking at the second texture, the shader boosts the brightness of the white areas and dims the black areas on the first texture. It does this relative to screen-space so the blinds pattern does not follow the sprite if they move (simulating the idea that the window sunlight/shadow is only being cast in one area of the room)

In theory, the best way to do this would be:

```renpy
label start:
    show eileen at window_shader

transform window_shader:
    shader "renpy.my_window_shader"
    tex1 "window_control_image.png"
```

with the idea that `tex1 "window_control_image.png"` would pass the provided image as a texture to the `tex1` variable in the shader.

Currently the only way I know of to use multiple textures in a displayable with a shader is to use the Model class and add them myself. However, this is not suitable for the above situation, because the first texture (the sprite) may be constantly changing expressions, so it isn't practical nor convenient to call some function to put together the sprite with the texture, and it also foregoes a lot of Ren'Py's conveniences like replacing expressions (e.g. `show eileen happy formal_outfit` followed by `show eileen sad` would retain the `formal_outfit` attribute, but if I were to write a function to manually put together the textures, I would need to either fetch or remember the attributes which are already on the sprite).

In short: is there a way of adding textures to transforms that have a shader? If so, it needs to be documented, and if not, I would like to request it as a feature.

**Comment by mal** [comment]
It is covered here: <https://www.renpy.org/doc/html/model.html#model-displayable>

I believe you'd want something like this (untested).
```renpy
transform window_shader(child=None):
    Model().texture(child).child('window_control_image.png')
    shader 'renpy.my_window_shader'
```

**Comment by shawna-p** [comment]
This *sort of* does what I need, though I will need to play around with it more to confirm. That said, I would suggest this is made clearer in the documentation, perhaps with a link to https://www.renpy.org/doc/html/atl.html#special-atl-keyword-parameters and a note on how to add new textures, as this question has come up before in the Ren'Py server and I wasn't able to identify that part of the page as containing the information I needed. I also noticed that the transform doesn't "keep" the child - 
```renpy
show eileen at window_shader
e "I am being shaded"
show eileen at left
``` 
gives me an error that tex1 has no provided value. It seems odd to me that the transform appears to persist on the sprite, yet it doesn't receive the child/texture unless I change it to `show eileen at left, window_shader`

**Comment by Gouvernathor** [comment]
I think that last part is intended behavior, see https://www.renpy.org/dev-doc/html/atl.html#replacing-transforms
The left transform replaces the window_shader transform, so the child override done by window_shader disappears (using Model() instead of pure eileen), but the `shader` property isnt, because `left` has no `shader None` property (and if it did it would break `show eileen at left, window_shader`).

Adding a `tex1`, `tex2` and so forth properties (or `gl_tex1`) could be considered, it would solve the merging issue and wouldn't be a bad idea in my opinion, but I don't know of a property taking a *displayable*, which is considerably harder to manage than a bunch of innocent floats.

Anyway, to solve your particular problem, I'd suggest `show expression At("eileen", window_shader) as eileen`, I think `show eileen at left` would work as you intend afterwards.

**Comment by LeyKlussyn** [comment]
Hi, I've run into a very similar problem, and I wanted to bump the issue because I think I have an interesting use case. 

I've joined a project that use shaders to emulate an old-school computer screen. Therefore, the shader run on the UI/screens layer. There is an original shader that worked fine, but when I made a new one, it started breaking things, especially buttons (quick menu) and all interactive elements more generally.

I've quickly found out the reason seems to be that the first one use `transform: mesh True`, and mine use `Model()`. However, in my shader, I can't use `mesh`, because **it works with noise textures as inputs.** And as the above shows, there's no way currently to add several textures to a shader transform. The only supported way is `Model()`.

In case a demo is necessary, here a short example:

```py
init python:
  renpy.register_shader("example01", variables="""
          uniform sampler2D tex0;
          uniform float u_lod_bias;
          attribute vec2 a_tex_coord;
          varying vec2 v_tex_coord;
      """,  vertex_700="""
          v_tex_coord = a_tex_coord;
      """, fragment_700="""
          gl_FragColor = texture2D(tex0, v_tex_coord.xy, u_lod_bias);
          gl_FragColor.rgb =  gl_FragColor.gbr;
      """)
# example 01 transform, using mesh. Works fine with screens:
transform shader01:
    mesh True
    shader 'example01'
# example 01 transform, using model(). Make buttons unusable:
init python:
    def shader01_model(d):
        model = Model().shader("example01").child(d, fit=True)
        return model
```
Applying `shader01` over a screen with a button doesn't break it, but `shader01_model` does. With an hypothetical "example02" shader that use tex0 and tex1, currently, it's not possible to write a similar `mesh` transform that wouldn't break the screen/button. I personally think a `gl_tex1` property would be very intuitive to use, but it's probably not as easy to implement.

**Comment by EtchJetty** [comment]
This would be greatly appreciated, especially in the use-case of effecting the `trans` object in functions.
Right now there's currently no way to transform `tex0` or `tex1` directly using Python that I know of. In an ideal situation one would be able to write a function like:

```py
def transformation_func(self, trans, st, at=None):
        trans.shader = 'hjt.fnafroom'
        trans.mesh = True
        trans.u_scroll = self.scroll
        trans.tex1 = "displinv"
```

**Comment by JamieDerg** [comment]
Has there been any Activity on this?

**Comment by renpytom** [comment]
This has been added in 8.4.0
